File: src/core/validate.ts
export interface WorkflowDefinition {
  definition: {
    $schema?: string;
    contentVersion?: string;
    triggers: Record<string, TriggerAction>;
    actions: Record<string, WorkflowAction>;
    parameters?: Record<string, unknown>;
    staticResults?: Record<string, unknown>;
  };
}

export interface TriggerAction {
  type?: string;
  kind?: string;
  inputs?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface WorkflowAction {
  type?: string;
  inputs?: unknown;
  runAfter?: Record<string, string[]>;
  expression?: unknown;
  actions?: Record<string, WorkflowAction>;
  else?: { actions: Record<string, WorkflowAction> };
  foreach?: string;
  [key: string]: unknown;
}

export function validateJson(output: string): WorkflowDefinition {
  if (typeof output !== 'string' || output.trim() === '') {
    throw new Error('Output is empty or not a string');
  }

  let cleaned = output.trim();

  // Strategy 1: Strip markdown code fences (anywhere in the output, not just wrapping it)
  const fenceMatch = cleaned.match(/```(?:json)?\s*\n?([\s\S]*?)\n?\s*```/);
  if (fenceMatch) {
    cleaned = fenceMatch[1].trim();
  }

  // Strategy 2: If it still doesn't start with {, extract the first { ... } block
  if (!cleaned.startsWith('{')) {
    const startIdx = cleaned.indexOf('{');
    const endIdx = cleaned.lastIndexOf('}');
    if (startIdx !== -1 && endIdx > startIdx) {
      cleaned = cleaned.slice(startIdx, endIdx + 1);
    }
  }

  let parsed: WorkflowDefinition;
  try {
    parsed = JSON.parse(cleaned) as WorkflowDefinition;
  } catch {
    throw new Error('Output is not valid JSON');
  }

  if (!parsed.definition || typeof parsed.definition !== 'object') {
    throw new Error('Missing required "definition" property');
  }

  if (!parsed.definition.actions || typeof parsed.definition.actions !== 'object') {
    throw new Error('Missing required "definition.actions" property');
  }

  return parsed;
}

/**
 * Deep structural validation of a Logic Apps workflow definition.
 * Returns an array of warning/error strings. Empty array means valid.
 */
export function validateWorkflowStructure(parsed: WorkflowDefinition): string[] {
  const issues: string[] = [];
  const def = parsed.definition;

  // Check triggers
  if (def.triggers && typeof def.triggers === 'object') {
    for (const [name, trigger] of Object.entries(def.triggers)) {
      if (!trigger.type || typeof trigger.type !== 'string') {
        issues.push(`Trigger "${name}" is missing a "type" field`);
      }
    }
  }

  // Check actions
  const actionNames = new Set(Object.keys(def.actions));

  for (const [name, action] of Object.entries(def.actions)) {
    if (!action.type || typeof action.type !== 'string') {
      issues.push(`Action "${name}" is missing a "type" field`);
    }

    // Validate runAfter references point to existing actions
    if (action.runAfter && typeof action.runAfter === 'object') {
      for (const dep of Object.keys(action.runAfter)) {
        if (!actionNames.has(dep)) {
          issues.push(`Action "${name}" has runAfter reference to non-existent action "${dep}"`);
        }
      }
    }

    // Condition actions must have If/Else structure
    if (action.type === 'If') {
      if (!action.expression) {
        issues.push(`Condition action "${name}" is missing "expression"`);
      }
      if (!action.actions && !action.else) {
        issues.push(`Condition action "${name}" has no "actions" or "else" branches`);
      }
    }

    // Foreach must have forEach and actions
    if (action.type === 'Foreach') {
      if (!action.foreach) {
        issues.push(`Foreach action "${name}" is missing "foreach" input`);
      }
      if (!action.actions || typeof action.actions !== 'object') {
        issues.push(`Foreach action "${name}" is missing nested "actions"`);
      }
    }
  }

  return issues;
}

File: src/mule2logic_agent/core/validate.py
"""JSON validation — ensures agent output is valid Logic Apps JSON."""

from __future__ import annotations

import json
import re
from typing import Any

from mule2logic_agent.models import WorkflowDefinition


def validate_json(output: str) -> WorkflowDefinition:
    """Parse and validate agent output as Logic Apps workflow JSON.

    Applies multiple strategies to extract valid JSON:
      1. Strip markdown code fences.
      2. Extract the first ``{ … }`` block if leading text is present.

    Raises ``ValueError`` on any structural validation failure.
    """
    if not isinstance(output, str) or not output.strip():
        raise ValueError("Output is empty or not a string")

    cleaned = output.strip()

    # Strategy 1: strip markdown code fences
    fence_match = re.search(r"```(?:json)?\s*\n?([\s\S]*?)\n?\s*```", cleaned)
    if fence_match:
        cleaned = fence_match.group(1).strip()

    # Strategy 2: extract the first { … } block
    if not cleaned.startswith("{"):
        start = cleaned.find("{")
        end = cleaned.rfind("}")
        if start != -1 and end > start:
            cleaned = cleaned[start : end + 1]

    try:
        parsed: dict[str, Any] = json.loads(cleaned)
    except json.JSONDecodeError as exc:
        raise ValueError("Output is not valid JSON") from exc

    if not isinstance(parsed.get("definition"), dict):
        raise ValueError('Missing required "definition" property')

    if not isinstance(parsed["definition"].get("actions"), dict):
        raise ValueError('Missing required "definition.actions" property')

    return parsed


def validate_workflow_structure(parsed: WorkflowDefinition) -> list[str]:
    """Deep structural validation of a Logic Apps workflow definition.

    Returns a list of warning/error strings.  Empty list means valid.
    """
    issues: list[str] = []
    definition = parsed["definition"]

    # Check triggers
    triggers = definition.get("triggers")
    if isinstance(triggers, dict):
        for name, trigger in triggers.items():
            if not isinstance(trigger, dict) or not isinstance(
                trigger.get("type"), str
            ):
                issues.append(f'Trigger "{name}" is missing a "type" field')

    # Check actions
    actions: dict[str, Any] = definition["actions"]
    action_names = set(actions.keys())

    for name, action in actions.items():
        if not isinstance(action, dict):
            continue

        if not isinstance(action.get("type"), str):
            issues.append(f'Action "{name}" is missing a "type" field')

        # Validate runAfter references
        run_after = action.get("runAfter")
        if isinstance(run_after, dict):
            for dep in run_after:
                if dep not in action_names:
                    issues.append(
                        f'Action "{name}" has runAfter reference to '
                        f'non-existent action "{dep}"'
                    )

        # Condition (If) validation
        if action.get("type") == "If":
            if not action.get("expression"):
                issues.append(
                    f'Condition action "{name}" is missing "expression"'
                )
            if not action.get("actions") and not action.get("else"):
                issues.append(
                    f'Condition action "{name}" has no "actions" or "else" branches'
                )

        # Foreach validation
        if action.get("type") == "Foreach":
            if not action.get("foreach"):
                issues.append(
                    f'Foreach action "{name}" is missing "foreach" input'
                )
            if not isinstance(action.get("actions"), dict):
                issues.append(
                    f'Foreach action "{name}" is missing nested "actions"'
                )

    return issues
